File: crates/brainwires-mcp-server/src/middleware/rate_limit.rs
use async_trait::async_trait;
use brainwires_mcp_client::{JsonRpcError, JsonRpcRequest};
use std::collections::HashMap;
use std::sync::Arc;
use tokio::sync::Mutex;
use tokio::time::Instant;

use super::{Middleware, MiddlewareResult};
use crate::connection::RequestContext;

struct RateLimitBucket {
    tokens: f64,
    last_refill: Instant,
}

/// Token-bucket rate limiting middleware.
pub struct RateLimitMiddleware {
    max_requests_per_second: f64,
    per_tool_limits: HashMap<String, f64>,
    buckets: Arc<Mutex<HashMap<String, RateLimitBucket>>>,
}

impl RateLimitMiddleware {
    /// Create a new rate limiter with a global requests-per-second limit.
    pub fn new(max_requests_per_second: f64) -> Self {
        Self {
            max_requests_per_second,
            per_tool_limits: HashMap::new(),
            buckets: Arc::new(Mutex::new(HashMap::new())),
        }
    }

    /// Set a per-tool rate limit override.
    pub fn with_tool_limit(mut self, tool_name: &str, limit: f64) -> Self {
        self.per_tool_limits.insert(tool_name.to_string(), limit);
        self
    }

    fn get_limit(&self, key: &str) -> f64 {
        self.per_tool_limits
            .get(key)
            .copied()
            .unwrap_or(self.max_requests_per_second)
    }
}

#[async_trait]
impl Middleware for RateLimitMiddleware {
    async fn process_request(
        &self,
        request: &JsonRpcRequest,
        _ctx: &mut RequestContext,
    ) -> MiddlewareResult {
        // Only rate-limit tools/call
        if request.method != "tools/call" {
            return MiddlewareResult::Continue;
        }

        let tool_name = request
            .params
            .as_ref()
            .and_then(|p| p.get("name"))
            .and_then(|n| n.as_str())
            .unwrap_or("unknown");

        let limit = self.get_limit(tool_name);
        let key = format!("tool:{tool_name}");

        let mut buckets = self.buckets.lock().await;
        let bucket = buckets.entry(key).or_insert(RateLimitBucket {
            tokens: limit,
            last_refill: Instant::now(),
        });

        // Token bucket refill
        let now = Instant::now();
        let elapsed = now.duration_since(bucket.last_refill).as_secs_f64();
        bucket.tokens = (bucket.tokens + elapsed * limit).min(limit);
        bucket.last_refill = now;

        if bucket.tokens >= 1.0 {
            bucket.tokens -= 1.0;
            MiddlewareResult::Continue
        } else {
            MiddlewareResult::Reject(JsonRpcError {
                code: -32002,
                message: format!("Rate limited: too many requests for tool '{tool_name}'"),
                data: None,
            })
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::connection::RequestContext;
    use serde_json::json;

    fn tools_call_request(tool_name: &str) -> JsonRpcRequest {
        JsonRpcRequest {
            jsonrpc: "2.0".to_string(),
            id: json!(1),
            method: "tools/call".to_string(),
            params: Some(json!({"name": tool_name})),
        }
    }

    fn non_tool_request() -> JsonRpcRequest {
        JsonRpcRequest {
            jsonrpc: "2.0".to_string(),
            id: json!(1),
            method: "tools/list".to_string(),
            params: None,
        }
    }

    #[tokio::test]
    async fn non_tool_call_passes_rate_limiter() {
        let middleware = RateLimitMiddleware::new(1.0);
        let req = non_tool_request();
        let mut ctx = RequestContext::new(json!(1));
        let result = middleware.process_request(&req, &mut ctx).await;
        assert!(matches!(result, MiddlewareResult::Continue));
    }

    #[tokio::test]
    async fn first_tool_call_passes() {
        let middleware = RateLimitMiddleware::new(10.0);
        let req = tools_call_request("my_tool");
        let mut ctx = RequestContext::new(json!(1));
        let result = middleware.process_request(&req, &mut ctx).await;
        assert!(matches!(result, MiddlewareResult::Continue));
    }

    #[tokio::test]
    async fn zero_rate_limit_rejects_immediately() {
        // A limit of 0 means no tokens are ever available after the first call
        // Initial bucket starts with `limit` tokens, so first call with limit=0 has 0 tokens
        let middleware = RateLimitMiddleware::new(0.0);
        let req = tools_call_request("slow_tool");
        let mut ctx = RequestContext::new(json!(1));
        let result = middleware.process_request(&req, &mut ctx).await;
        assert!(matches!(result, MiddlewareResult::Reject(ref e) if e.code == -32002));
    }

    #[tokio::test]
    async fn per_tool_limit_override_applied() {
        let middleware = RateLimitMiddleware::new(100.0).with_tool_limit("special_tool", 0.0);
        let req = tools_call_request("special_tool");
        let mut ctx = RequestContext::new(json!(1));
        let result = middleware.process_request(&req, &mut ctx).await;
        assert!(matches!(result, MiddlewareResult::Reject(_)));
    }

    #[test]
    fn get_limit_uses_default_when_no_override() {
        let middleware = RateLimitMiddleware::new(5.0);
        assert_eq!(middleware.get_limit("any_tool"), 5.0);
    }

    #[test]
    fn get_limit_uses_override_when_set() {
        let middleware = RateLimitMiddleware::new(5.0).with_tool_limit("special", 2.0);
        assert_eq!(middleware.get_limit("special"), 2.0);
        assert_eq!(middleware.get_limit("other"), 5.0);
    }
}

File: deno/packages/mcp-server/middleware/rate_limit.ts
/**
 * @module middleware/rate_limit
 *
 * Token-bucket rate limiting middleware.
 * Equivalent to Rust's `RateLimitMiddleware`.
 */

import type { JsonRpcRequest } from "@rullama/mcp-client";
import type { RequestContext } from "../server.ts";
import {
  type Middleware,
  middlewareContinue,
  middlewareReject,
  type MiddlewareResult,
} from "./mod.ts";

interface RateLimitBucket {
  tokens: number;
  lastRefill: number; // timestamp ms
}

/** Buckets kept before the least-recently-used one is dropped. */
export const MAX_RATE_LIMIT_BUCKETS = 1024;
/** Longest tool name that gets its own bucket; longer names share one. */
const MAX_KEY_TOOL_NAME = 128;

/**
 * Token-bucket rate limiting middleware.
 * Only rate-limits tools/call requests.
 * Equivalent to Rust `RateLimitMiddleware`.
 */
export class RateLimitMiddleware implements Middleware {
  private readonly maxRequestsPerSecond: number;
  private readonly perToolLimits: Map<string, number> = new Map();
  private readonly buckets: Map<string, RateLimitBucket> = new Map();

  constructor(maxRequestsPerSecond: number) {
    this.maxRequestsPerSecond = maxRequestsPerSecond;
  }

  /** Set a per-tool rate limit override. Returns this for chaining. */
  withToolLimit(toolName: string, limit: number): this {
    this.perToolLimits.set(toolName, limit);
    return this;
  }

  private getLimit(key: string): number {
    return this.perToolLimits.get(key) ?? this.maxRequestsPerSecond;
  }

  /** Get-or-create the bucket for `key`, refreshing its LRU position. */
  private bucketFor(key: string, limit: number): RateLimitBucket {
    const existing = this.buckets.get(key);
    if (existing) this.buckets.delete(key); // refresh recency
    const bucket = existing ?? { tokens: limit, lastRefill: Date.now() };
    this.buckets.set(key, bucket);
    while (this.buckets.size > MAX_RATE_LIMIT_BUCKETS) {
      this.buckets.delete(this.buckets.keys().next().value as string);
    }
    return bucket;
  }

  /** Refill by elapsed time, then take one token; false when none is left. */
  private static take(bucket: RateLimitBucket, limit: number): boolean {
    const now = Date.now();
    const elapsed = (now - bucket.lastRefill) / 1000;
    bucket.tokens = Math.min(bucket.tokens + elapsed * limit, limit);
    bucket.lastRefill = now;
    if (bucket.tokens < 1.0) return false;
    bucket.tokens -= 1.0;
    return true;
  }

  processRequest(
    request: JsonRpcRequest,
    ctx: RequestContext,
  ): Promise<MiddlewareResult> {
    // Only rate-limit tools/call
    if (request.method !== "tools/call") {
      return Promise.resolve(middlewareContinue());
    }
    const params = request.params as Record<string, unknown> | undefined;
    const rawName = params?.name;
    const toolName =
      typeof rawName === "string" && rawName.length <= MAX_KEY_TOOL_NAME
        ? rawName
        : "unknown";
    const limit = this.getLimit(toolName);
    // Keyed per client so one client cannot exhaust another's budget. The map
    // is bounded (LRU) because the tool name is attacker-controlled input.
    const key = `${ctx.clientInfo?.name ?? "anonymous"}:${toolName}`;
    if (RateLimitMiddleware.take(this.bucketFor(key, limit), limit)) {
      return Promise.resolve(middlewareContinue());
    }
    return Promise.resolve(middlewareReject({
      code: -32002,
      message: `Rate limited: too many requests for tool '${toolName}'`,
    }));
  }
}
